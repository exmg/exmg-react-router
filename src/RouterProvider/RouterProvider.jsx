import { Component } from 'react';
import PropTypes from 'prop-types';
import createHistory from 'history/createHashHistory';

export const routerContextTypes = {
	router: PropTypes.shape({
		path: PropTypes.string.isRequired,
		register: PropTypes.func.isRequired,
		history: PropTypes.shape({
			push: PropTypes.func.isRequired,
			replace: PropTypes.func.isRequired,
		}).isRequired,
	}),
};

export default class RouterProvider extends Component {
	static propTypes = {
		children: PropTypes.node,
	};

	static defaultProps = {
		children: null,
	};

	static childContextTypes = routerContextTypes;

	// eslint-disable-next-line react/sort-comp
	history = createHistory();
	routes = [];
	routeMap = {};

	getChildContext() {
		return {
			router: {
				// Pass path to build op child routes
				path: '',
				register: this.register,
				history: this.history,
			},
		};
	}

	componentWillMount() {
		this.unlisten = this.history.listen(this.updateAction);
	}

	componentDidMount() {
		this.updateAction();
	}

	componentWillUnmount() {
		this.unlisten();
	}

	register = (route) => {
		this.routes.push(route);

		const parentPath = route.context.router.path;

		if (!this.routeMap[parentPath]) {
			this.routeMap[parentPath] = [route];
		} else {
			this.routeMap[parentPath].push(route);
		}

		const routeMap = this.routeMap[parentPath];

		this.requestUpdate();

		return () => {
			const index = routeMap.indexOf(route);

			routeMap.splice(index, 1);
		};
	}

	requestUpdate() {
		if (!this.rAF) {
			this.rAF = requestAnimationFrame(this.updateAction);
		}
	}

	updateAction = () => {
		const { pathname } = this.history.location;

		Object.values(this.routeMap).forEach((routes) => {
			let foundMatch = false;
			const nonMatchingRoutes = [];

			routes.forEach((route) => {
				const match = route.match(pathname);

				if (!route.props.notFound) {
					route.update(match);
				}

				if (route.props.exclude) {
					return;
				}

				if (match) {
					foundMatch = true;
				}

				if (route.props.notFound) {
					nonMatchingRoutes.push(route);
				}
			});

			if (nonMatchingRoutes.length) {
				nonMatchingRoutes.forEach(route => route.update(!foundMatch));
			}
		});

		this.rAF = null;
	}

	render() {
		return this.props.children;
	}
}
