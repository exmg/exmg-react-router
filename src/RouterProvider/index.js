export { default as RouterProvider, routerContextTypes } from './RouterProvider';
