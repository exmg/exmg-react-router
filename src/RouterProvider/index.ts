export { default as RouterProvider } from './RouterProvider';
