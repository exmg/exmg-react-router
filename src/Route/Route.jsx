import { Component } from 'react';
import PropTypes from 'prop-types';
import pathToRegexp from 'path-to-regexp';

import { routerContextTypes } from '../RouterProvider/RouterProvider';
import { shallowEqual, routerStateEqual } from '../util';

export default class Route extends Component {
	static propTypes = {
		children: PropTypes.oneOfType([
			PropTypes.node,
			PropTypes.func,
		]),
		exact: PropTypes.bool,
		path: PropTypes.string,
		notFound: PropTypes.bool, // noMatch ?
		// eslint-disable-next-line react/no-unused-prop-types
		exclude: PropTypes.bool, // exclude from route notFound handling
	};

	static defaultProps = {
		children: null,
		exact: false,
		notFound: false,
		path: '/',
		exclude: false,
	};

	static contextTypes = routerContextTypes;
	static childContextTypes = routerContextTypes;

	state = {
		match: false,
	};

	// eslint-disable-next-line react/sort-comp
	keys = [];
	pathRegexp = pathToRegexp(this.path, this.keys, { end: false });

	getChildContext() {
		return {
			router: {
				...this.context.router,
				path: this.path,
			},
		};
	}

	componentDidMount() {
		const { router } = this.context;

		this.unregister = router.register(this);
	}

	shouldComponentUpdate(nextProps, nextState) {
		const propsChanged = !shallowEqual(this.props, nextProps);
		const stateChanged = !routerStateEqual(this.state, nextState);

		return propsChanged || stateChanged;
	}

	componentWillUnmount() {
		this.unregister();
	}

	get path() {
		const path = this.context.router.path + this.props.path;

		return path
			.replace(/\/\/|\/$/g, '/')
			.replace(/\/+$/, '');
	}

	match(pathname) {
		if (this.props.notFound) {
			return false;
		}

		const match = this.pathRegexp.exec(pathname);

		if (!match) {
			return false;
		}

		const [url, ...values] = match;
		const isExact = pathname === url;

		if (this.props.exact && !isExact) {
			return false;
		}

		return {
			pathname,
			params: this.keys.reduce((memo, key, index) => {
				// eslint-disable-next-line no-param-reassign
				memo[key.name] = values[index];
				return memo;
			}, {}),
		};
	}

	update(match) {
		this.setState({
			match,
		});
	}

	render() {
		const { children } = this.props;
		const { match } = this.state;

		if (typeof children === 'function') {
			// Empty object so we can always destruct
			return children(!!match, match ? match.params : {});
		}

		if (match) {
			return children;
		}

		return null;
	}
}
