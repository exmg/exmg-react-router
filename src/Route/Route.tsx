import React, { Component } from 'react';
import pathToRegexp from 'path-to-regexp';

import { Context } from '../RouterProvider';

type ChildrenFunc = (match: boolean, params?: Params) => React.ReactNode;

export interface Props {
  children: React.ReactNode | ChildrenFunc;
  exact?: boolean;
  path?: string;
  notFound?: boolean;
  exclude?: boolean;
}

export interface Params {
  [key: string]: any;
}

export interface Match {
  params: Params;
  pathname: string;
}

export interface State {
  match: false | Match;
}

export class RouteComponent extends Component<Props & Context, State> {
  static defaultProps = {
    path: '/',
    exact: false,
    notFound: false,
    exclude: false,
  };

  state = {
    match: false,
  } as State;

  keys: pathToRegexp.Key[] = [];
  pathRegexp = pathToRegexp(this.path, this.keys, { end: false });
  unregister: () => void;

  get path() {
    const path = this.props.parentPath + this.props.path;

    return path
			.replace(/\/\/|\/$/g, '/')
			.replace(/\/+$/, '');
  }

  componentDidMount() {
    this.unregister = this.props.register(this);
  }

  componentWillUnmount() {
    this.unregister();
  }

  match(pathname: string): false | Match {
    if (this.props.notFound) {
      return false;
    }

    const match = this.pathRegexp.exec(pathname);

    if (!match) {
      return false;
    }

    const [matchPath, ...values] = match;

    if (this.props.exact && pathname !== matchPath) {
      return false;
    }

    const params: Params = {};

    this.keys.forEach((key, index) => {
      params[key.name] = values[index];
    });

    return {
      pathname,
      params,
    };
  }

  update(match: State['match']) {
    this.setState({
      match,
    });
  }

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    if (this.state.match === nextState.match && this.props === nextProps) {
      return false;
    }

    return true;
  }

  render() {
    const { children } = this.props;
    const { match } = this.state;

    if (typeof children === 'function') {
      if (match) {
        return children(true, match.params);
      }

      return children(false, {});
    }

    if (match) {
      return children;
    }

    return null;
  }
}

export class Route extends Component<Props> {
  // register: Context['register'] = (component) => {
  //   console.log('Route::register', component);
  // }

  // contextProps: Context = {
  //   register: this.register,
  //   parentPath: this.props.path,
  // };

  render() {
    const { children, ...props } = this.props;

    return (
      <Context.Consumer>
        { router => (
          <Context.Provider value={ { register: router.register, parentPath: this.props.path } }>
            <RouteComponent { ...router } { ...props }>
              { children }
            </RouteComponent>
          </Context.Provider>
        ) }
      </Context.Consumer>
    );
  }
}

// export const Route: React.SFC<Props> = ({ children, ...props }) => (
//   <Context.Consumer>
//     { router => (
//       <RouteComponent { ...router } { ...props }>
//         <Context.Provider value={ router }>
//           { children }
//         </Context.Provider>
//       </RouteComponent>
//     ) }
//   </Context.Consumer>
// );
