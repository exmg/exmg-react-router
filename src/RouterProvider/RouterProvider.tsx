import React, { Component } from 'react';

import { createHistory, HistoryTypes } from '../utils';
import { Route } from 'Route/Route';
import { UnregisterCallback  } from 'history';

export interface Context {
  register: (component: Component) => (() => void);
  parentPath: string;
}

export const Context = React.createContext<Context>({} as Context);

export interface Props {
  children: React.ReactNode;
  type: HistoryTypes;
}

interface RoutesMap {
  [key: string]: Route[];
}

export default class RouterProvider extends Component<Props> {
  static defaultProps = {
    type: 'hash',
  };

  raf?: number;
  routesMap: RoutesMap = {};
  unlisten: UnregisterCallback;
  history = createHistory(this.props.type);

  register: Context['register'] = (routeComponent: Route) => {
    const { parentPath } = routeComponent.context;

    if (!this.routesMap[parentPath]) {
      this.routesMap[parentPath] = [];
    }

    const map = this.routesMap[parentPath];

    map.push(routeComponent);

    this.requestUpdate();

    return () => {
      const index = map.indexOf(routeComponent);

      map.splice(index, 1);

      this.requestUpdate();
    };
  }

  contextProps: Context = {
    register: this.register,
    parentPath: '/',
  };

  componentDidMount() {
    this.unlisten = this.history.listen(this.update);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  requestUpdate() {
    if (!this.raf) {
      this.raf = requestAnimationFrame(this.update);
    }
  }

  update = () => {
    const { pathname } = this.history.location;
    const routeGroups = Object.values(this.routesMap);

    routeGroups.forEach(routes => this.updateGroup(pathname, routes));

    this.raf = null;
  }

  updateGroup(pathname: string, routes: Route[]) {
    let foundMatch = false;
    const nonMatchingRoutes: Route[] = [];

    routes.forEach((route) => {
      const match = route.match(pathname);

      if (!route.props.notFound) {
        route.update(match);
      }

      if (route.props.exclude) {
        return;
      }

      if (match) {
        foundMatch = true;
      }

      if (route.props.notFound) {
        nonMatchingRoutes.push(route);
      }
    });

    if (nonMatchingRoutes.length) {
      const notFoundMatch = foundMatch ? false : { pathname, params: {} };

      nonMatchingRoutes.forEach(route => route.update(notFoundMatch));
    }
  }

  render() {
    return (
      <Context.Provider value={ this.contextProps }>
        { this.props.children }
      </Context.Provider>
    );
  }
}
