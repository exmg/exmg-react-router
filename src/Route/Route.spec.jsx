import React from 'react';
import renderer from 'react-test-renderer';

import RouterProvider from '../RouterProvider/RouterProvider';
import Route from './Route';

global.requestAnimationFrame = cb => cb();

const setHash = hash => new Promise((resolve) => {
	window.location.hash = hash;

	setTimeout(resolve, 100);
});

test('Route renders children', (done) => {
	setHash('/').then(() => {
		const component = renderer.create(
			<RouterProvider>
				<Route path="/" exact>Home</Route>
				<Route path="/foo">Foo</Route>
			</RouterProvider>,
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		done();
	});
});

test('Route renders functions', (done) => {
	setHash('/').then(() => {
		const component = renderer.create(
			<RouterProvider>
				<Route path="/" exact>{match => match && 'Foo'}</Route>
				<Route path="/foo">{match => match && 'Bar'}</Route>
				<Route path="/foo/bar">{() => 'Render anyway'}</Route>
			</RouterProvider>,
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		done();
	});
});

test('Route renders functions route: /foo', (done) => {
	setHash('/foo').then(() => {
		const component = renderer.create(
			<RouterProvider>
				<Route path="/" exact>{match => match && 'Foo'}</Route>
				<Route path="/foo">{match => match && 'Bar'}</Route>
				<Route path="/foo/bar">{() => 'Render anyway'}</Route>
			</RouterProvider>,
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		done();
	});
});

test('Route contains params', (done) => {
	setHash('/16/barry/some/args').then(() => {
		const component = renderer.create(
			<RouterProvider>
				<Route path="/:id/:name/:rest*">{(match, params) => JSON.stringify(params)}</Route>
				<Route path="/:rest*">{(match, params) => JSON.stringify(params)}</Route>
			</RouterProvider>,
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		done();
	});
});

test('Renders notFound route', (done) => {
	setHash('/404').then(() => {
		const component = renderer.create(
			<RouterProvider>
				<Route path="/" exact>Home</Route>
				<Route path="/bar">Bar</Route>
				<Route notFound>404 Not Found</Route>
			</RouterProvider>,
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		done();
	});
});
