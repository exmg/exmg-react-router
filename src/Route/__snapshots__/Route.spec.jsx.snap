// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders notFound route 1`] = `"404 Not Found"`;

exports[`Route contains params 1`] = `
Array [
  "{\\"id\\":\\"16\\",\\"name\\":\\"barry\\",\\"rest\\":\\"some/args\\"}",
  "{\\"rest\\":\\"16/barry/some/args\\"}",
]
`;

exports[`Route renders children 1`] = `"Home"`;

exports[`Route renders functions 1`] = `
Array [
  "Foo",
  "Render anyway",
]
`;

exports[`Route renders functions route: /foo 1`] = `
Array [
  "Bar",
  "Render anyway",
]
`;
